jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/CreateLambdaFunction.kt

```diff
@@ -6,26 +6,58 @@ package software.aws.toolkits.jetbrains.services.lambda.upload
 import com.intellij.openapi.actionSystem.AnAction
 import com.intellij.openapi.actionSystem.AnActionEvent
 import com.intellij.openapi.actionSystem.LangDataKeys
+import com.intellij.psi.PsiElement
 import icons.AwsIcons
 import software.amazon.awssdk.services.lambda.model.Runtime
+import software.aws.toolkits.jetbrains.services.cloudformation.CloudFormationTemplateIndex
 import software.aws.toolkits.jetbrains.services.iam.IamRole
+import software.aws.toolkits.jetbrains.services.lambda.LambdaHandlerResolver
 import software.aws.toolkits.jetbrains.services.lambda.runtime
 import software.aws.toolkits.jetbrains.services.lambda.upload.EditFunctionMode.NEW
 import software.aws.toolkits.resources.message
 
-class CreateLambdaFunction(private val handlerName: String? = null) : AnAction(message("lambda.create_new"), null, AwsIcons.Actions.LAMBDA_FUNCTION_NEW) {
+class CreateLambdaFunction(
+    private val handlerName: String?,
+    private val element: PsiElement?,
+    private val lambdaHandlerResolver: LambdaHandlerResolver?
+) : AnAction(message("lambda.create_new"), null, AwsIcons.Actions.LAMBDA_FUNCTION_NEW) {
+
+    init {
+        if (handlerName != null) {
+            element ?: throw Error("element must be provided if handlerName is provided")
+            lambdaHandlerResolver ?: throw Error("lambdaHandlerResolver must be provided if handlerName is provided")
+        }
+    }
+
     override fun actionPerformed(event: AnActionEvent) {
         val project = event.getRequiredData(LangDataKeys.PROJECT)
         val runtime = event.runtime()
 
-        val dialog = if (handlerName != null) {
-            EditFunctionDialog(project = project, mode = NEW, runtime = runtime, handlerName = handlerName)
-        } else {
+        val dialog = if (handlerName == null) {
             EditFunctionDialog(project = project, mode = NEW, runtime = runtime)
+        } else {
+            EditFunctionDialog(project = project, mode = NEW, runtime = runtime, handlerName = handlerName)
         }
 
         dialog.show()
     }
+
+    override fun update(e: AnActionEvent?) {
+        super.update(e)
+
+        if (handlerName == null || element == null || lambdaHandlerResolver == null) {
```

Comment:
since `handlerName` being non-null forces the other 2 to be non-null also do you need to include them in the if statement?

Based on `init` it's not possible to construct this class with `handlerName` being non-null but `element` or `lambdaHandlerResolver` is null.

That said - I guess one thing that this does give you is the smart casting... so maybe forget me ?

Reply:
Smart casting and defensive programming by not making assumptions on the init validation.

Reply:
kk